xdem/spstats.py

```diff
@@ -9,19 +9,123 @@
 import random
 import warnings
 from functools import partial
-from typing import Callable, Union
+from typing import Callable, Union, Iterable, List, Optional, Sequence
 
+import itertools
 import matplotlib.pyplot as plt
+from numba import njit
 import numpy as np
 import pandas as pd
 from scipy import integrate
 from scipy.optimize import curve_fit
+from scipy.stats import binned_statistic, binned_statistic_2d, binned_statistic_dd
+from xdem.spatial_tools import nmad
 
 with warnings.catch_warnings():
     warnings.filterwarnings("ignore", category=DeprecationWarning)
     import skgstat as skg
     from skgstat import models
 
+def nd_binning_scipy(dh: np.ndarray, list_var: List[np.ndarray], list_var_names=List[str], list_var_bins: Optional[List[Iterable]] = None,
+                     statistics: list[Union[str, Callable, None]] = ['count', np.nanmedian ,nmad], list_ranges : Optional[List[Sequence]] = None) \
+        -> tuple[list[pd.DataFrame],list[pd.DataFrame],pd.DataFrame]:
+    """
+    Quantify non-stationarities in elevation measurement errors with one or several, terrain or instrument-related,
+    explanatory variables. See documentation of scipy.stats.binned_statistic_dd for more details.
+
+    :param dh: elevation differences
+    :param list_var: list of explanatory variables
+    :param list_var_names: list of names of the explanatory variables
+    :param list_var_bins: list of bins for the explanatory variables
+    :param statistics: list of statistics to be computed
+    :param list_ranges: list of minimum and maximum ranges to bin the explanatory variables
+    :return:
+    """
+
+    # we separate 1d, 2d and nd binning, because propagating statistics between different dimensional binning is not always feasible
+    # using scipy because it allows for several dimensional binning, while it's not straightforward in pandas
+    if list_var_bins is None:
+        list_var_bins = (10,) * len(list_var_names)
+    statistics_name = [f if isinstance(f,str) else f.__name__ for f in statistics]
+
+    # get binned statistics in 1d: a simple loop is sufficient
+    list_df_1d = []
+    for i, var in enumerate(list_var):
+        df_stats_1d = pd.DataFrame()
+        # get statistics
+        for j, statistic in enumerate(statistics):
+            stats_binned_1d, bedges_1d = binned_statistic(var,dh,statistic=statistic,bins=list_var_bins[i],range=list_ranges)[:2]
+            # save in a dataframe
+            df_stats_1d[statistics_name[j]] = stats_binned_1d
+        # we need to get the middle of the bins from the edges, to get the same dimension length
+        df_stats_1d['bin_mid_var1'] = (bedges_1d[1:] + bedges_1d[:-1]) / 2
```

Comment:
I think Erik uses a magic panda function to do this autmoatically in the hypsometric_binning. See https://github.com/GlacioHack/xdem/blob/2a389a09ae4d68f7e765e576b1c9afd99021dc92/xdem/volume.py#L88

Reply:
So the index will contain the min, max and mid value if I'm correct.

Reply:
Modified the bins using Interval Indexes, and extending those to N-D in 985722f

Reply:
Yes, IntervalIndex is great! It has `left`, `right` and `mid` properties. I guess it's not called min/max because they might be either ascending or descending.